packages/notifications/src/notifications-provider.ts

```diff
@@ -0,0 +1,82 @@
+import { AnyPromiseFn } from '@terra-ui-packages/core-utils';
+
+import { withErrorHandling } from './error';
+
+/**
+ * Should return true of a given error should be ignored
+ */
+export type IgnoreErrorDecider = (title: string, obj?: unknown) => boolean;
+
+export type NotificationType = 'error' | 'warn' | 'info' | 'success' | 'welcome';
+export interface NotificationOptions {
+  /**
+   * string, Error(unknown), or json object to be displayed in detail section
+   */
+  detail?: unknown;
+}
+
+export interface Notifier {
+  notify: (type: NotificationType, title: string, options?: NotificationOptions) => void;
+}
+
+export interface ErrorReportingOptions {
+  rethrow?: boolean; // default: false
+  onReported?: () => void;
+}
+
+export interface ErrorReporter {
+  /**
+   * Reports the error visually to the user using the app's notification system
+   * @param title - error title
+   * @param obj - an error, response, or arbitrary json-style object
+   */
+  reportError: (title: string, obj?: unknown) => Promise<void>;
+
+  /**
+   * Returns a function augmenter (HoF).
+   * If provided function fails when called, report the error to the user with `title` as a side effect.
+   * Use options arg to give onReported callback, and/or rethrow on error.
+   */
+  withErrorReporting: <F extends AnyPromiseFn>(title: string, options?: ErrorReportingOptions) => (fn: F) => F;
+}
+
+export interface NotificationsProvider extends Notifier, ErrorReporter {}
+
+/**
+ * dependencies for NotificationsProvider factory method makeNotificationsProvider
+ */
+export interface NotificationsProviderDeps {
+  notifier: Notifier;
+  shouldIgnoreError: IgnoreErrorDecider;
+}
+
+/**
+ * allows composition of NotificationsProvider dependency when useful for code reuse
+ * @param deps
+ */
+export const makeNotificationsProvider = (deps: NotificationsProviderDeps): NotificationsProvider => {
```

Comment:
For the record, my suggestion was actually that only the second argument was an options object (because it contained _optional_ arguments).

`notifier` is always required, so it makes sense to be a positional argument.

Something like:

```suggestion
export const makeNotificationsProvider = (notifier: Notifier, options: NotificationsProviderOptions = {}): NotificationsProvider => {
  const { shouldIgnoreError = () => false } = options;
```

Reply:
yeah, I know, but I do like expressive arg objects in general... also, both args are actually required, tho I thought about making the 2nd one optional.